Weak etag

stokito opened the pull request:
In scope of #4277 and #4543 was added support of ETag.
But in the ` docs/cmdline-opts/etag-save.d ` file said:

```
 When server sends an
ETag, it must be enveloped by a double quote. This option extracts the
ETag without the double quotes and saves it into the <file>.

A server can send a week ETag which is prefixed by "W/". This identifier
is not considered, and only relevant ETag between quotation marks is parsed.
```

This is actually wrong behavior of client. According to spec client should only send the same ETag received from server and it's server's business to determine what to return 200 or 304 Not Modified.
Also now the code if fact converts weak etags to strong because it simply copy text from double quotes.
See https://tools.ietf.org/html/rfc7232#section-2.3.2 for details.
Fortunately weak ETag not so often used and the only web server that returns them for static files is Tomcat.
In fact this is a bug in Tomcat but we can use it as a test.

### I did this
Run the Tomcat with a `ololo.txt` file:
```
mkdir /tmp/www
echo "ololo" > /tmp/www/ololo.txt
docker run -it --rm --name tomcat -p 8080:8080 -v /tmp/www:/usr/local/tomcat/webapps/ROOT/ tomcat:latest
```
Fetch the file via curl built from sources:
```
$ /opt/bin/curl -s -vv http://localhost:8080/ololo.txt --etag-save ololo.txt.etag --etag-compare ololo.txt.etag
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /ololo.txt HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.72.0-DEV
> Accept: */*
> If-None-Match: ""
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 
< Accept-Ranges: bytes
< ETag: W/"6-1595194668152"
< Last-Modified: Sun, 19 Jul 2020 21:37:48 GMT
< Content-Type: text/plain
< Content-Length: 6
< Date: Sun, 19 Jul 2020 21:39:48 GMT
< 
ololo
* Connection #0 to host localhost left intact

$ cat ololo.txt.etag 
6-1595194668152

$ /opt/bin/curl -s -vv http://localhost:8080/ololo.txt --etag-save ololo.txt.etag --etag-compare ololo.txt.etag
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /ololo.txt HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.72.0-DEV
> Accept: */*
> If-None-Match: "6-1595194668152"
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 
< Accept-Ranges: bytes
< ETag: W/"6-1595194668152"
< Last-Modified: Sun, 19 Jul 2020 21:37:48 GMT
< Content-Type: text/plain
< Content-Length: 6
< Date: Sun, 19 Jul 2020 21:41:39 GMT
< 
ololo
* Connection #0 to host localhost left intact

```
As you may see at first request the `ololo.txt.etag` file not exists but  `If-None-Match: ""` was sent. It shouldn't be sent to minimize network load and not to confuse server.
The received ETag was weak ` W/"6-1595194668152"` but was saved as strong ETag `6-1595194668152`.
Next time the `6-1595194668152` was sent as a strong ETag `"6-1595194668152"` but server not recognized it and again returned 200 with payload and the same weak `ETag: W/"6-1595194668152"`.

### I expected the following
1. Empty Etag wasn't sent.
2. The returned ETag was saved "as is" to the file with `W/` prefix  and double quotes are not stripped.

### curl/libcurl version

    curl 7.72.0-DEV (x86_64-pc-linux-gnu) libcurl/7.72.0-DEV OpenSSL/1.1.1f zlib/1.2.11
    Release-Date: [unreleased]
    Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
    Features: AsynchDNS HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets


### Fix
Works :)
```
$ /opt/bin/curl -vv http://localhost:8080/ololo.txt --etag-save ololo.txt.etag --etag-compare ololo.txt.etag
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /ololo.txt HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.72.0-DEV
> Accept: */*
> If-None-Match: W/"6-1595194668152"
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 304 
< ETag: W/"6-1595194668152"
< Date: Sun, 19 Jul 2020 22:15:12 GMT
```

The PR is without tests because I have to sleep now 😴

@s-3ntinel sorry that I removed a lot of your code but please review the PR if you'll have a minute.
Thank you for your contribution.

jay commented:
See also #5626

/cc @sterchelen

stokito commented:
TBH the whole #5626 is useless. ETag for client is just an opaque header that just should be resend.

stokito commented:
Ok, I changed commits and now empty etag files will be ignored. Also I removed backward compatibility so now quotes  will be added anyway. This means that after curl update clients will receive 200 instead of 304 but that's not a big deal because clients will just simply update theirs etag files and everything will work next time.

sterchelen commented:
> TBH the whole #5626 is useless. ETag for a client is just an opaque header that just should be resend.

I am not sure to understand your point of view :thinking: ... either we don't care of the etag value and we send back a corrupt value without noticing the user OR we check/validate the format and notify the user that he can't count on it. 

Run `make checksrc`:
```
./tool_operate.c:924:15: warning: if with space (SPACEBEFOREPAREN)
             if (strlen(etag_from_file)) {
               ^
./tool_operate.c:942:13: warning: else after closing brace on same line (BRACEELSE)
             } else {
             ^
```

stokito commented:
If server returned this ETag then it expects to receive it back the same. In the same time client have nothing to do if the ETag is broken. The parsing logic adds complexity and wastes CPU.
Remember clould servers and Greta Thunberg asking for heat minimizing :)
I fixed formatting, sorry that didn't noticed it.

bagder commented:
I think @stokito's line of reasoning makes sense: whatever the server sends in the Etag header we can/should send back verbatim. So by totally ignoring the syntax we probably are better off rather than trying to make sure it adheres to any spec.

bagder commented:
~~~
./tool_operate.c:942:13: warning: else after closing brace on same line (BRACEELSE)
             }else {
             ^

stokito commented:
> ```
> ./tool_operate.c:942:13: warning: else after closing brace on same line (BRACEELSE)
>              }else {
>              ^
> ```

Guys, please add an `.editorconfig` file so my IDE can automatically switch to this projects code style.

bagder commented:
> Guys, please add an .editorconfig file so my IDE can automatically switch to this project's code style.

Feel free to submit one.

bagder commented:
@stokito: you still have code style issues here!

stokito commented:
@bagder just pushed. If something still wrong please fix yourself, I don't have much time today :( I also wanted to add a unit test but looks like can't do that

bagder commented:
> I also wanted to add a unit test but looks like can't do that

Why not?

bagder commented:
> If something still wrong please fix yourself

I trust you didn't run any test cases? All etags tests seem to fail now...

stokito commented:
I can't fix those tests.
I changed test339:
```
<file name="log/etag339">
"asdf"
</file>
```

But the new error occurred:
```
+++ log/check-generated 2020-07-28 00:42:07.169970979 +0300
@@ -1 +0,0 @@
-"asdf"[CR][LF]

 - abort tests
TESTDONE: 0 tests out of 1 reported OK: 0%
TESTFAIL: These test cases failed: 339 
TESTDONE: 1 tests were considered during 1 seconds.
```

bagder commented:
If the tests don't work, then the code isn't reliable.

bagder commented on src/tool_cb_hdr.c:
```diff
@@ -104,42 +104,17 @@ size_t tool_header_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
 
       /* get length of desired etag */
-      etag_length = (size_t)last - (size_t)first;
+      etag_length = (size_t)end - strlen("\r\n") - (size_t)etag_h;
```
You must not assume CRLF!

stokito replied:
this is specified in HTTP that headers are separated with CRLF. I don't know why eol was added to the input line, maybe this should be fixed on the upper level. If you know how to fix that please do.

bagder replied:
curl needs to handle headers ending with CRLF or only LF

stokito replied:
not sure what to answer here: the input string have CRLF and we need to strip it. If you know how to fix it gracefully then push a commit to the branch

bagder replied:
The input string will have CRLF or just LF without any CR and the code needs to handle both cases.

stokito replied:
ok, now I see. Still not clear why header may have LF but if you are sure that this may happen then we should strip it gracefully. But maybe this should be fixed on top level i.e. the input `ptr` should be passed into the func with already trimmed EOL

bagder replied:
It is used in the curl test suite and very common "in the wild" too.

> maybe this should be fixed on top level i.e. the input ptr should be passed into the func with already trimmed EOL

Possibly, but that change should not be done within this PR =)

stokito replied:
ok, I'll add a commit with EOL skip

bagder commented on src/tool_operate.c:
```diff
@@ -923,7 +923,7 @@ static CURLcode single_transfer(struct GlobalConfig *global,
             if (strlen(etag_from_file)) {
               char *header = NULL;
-              header = aprintf("If-None-Match: \"%s\"", etag_from_file);
+              header = aprintf("If-None-Match: %s", etag_from_file);
```
maybe these two lines could be combined? seems pointless to assign it to NULL just the line before it gets assigned again....

stokito replied:
sure, I just moved the line here not changed it keep diff clearer.

stokito replied:
Should I add a separate commit with merge of these two lines?

bagder replied:
On second thoughts, let's just leave it like you've done it.

bagder commented:
I tried amending the tests and that made them work in my build.

bagder commented:
Are you still working on this?

stokito commented:
Code style was fixed.
The test was failed on my PC and I have no idea why. But you said that:

> I tried amending the tests and that made them work in my build.

So I guess you can push that fixed test to the branch.

Also I answered here to your comments and it looks like no any special actions needed from my side. So I waitng while you press "Resolve conversation"

stokito commented:
I added few commits that should resolve your review notes.

bagder commented:
The tests 344 345 1566 are still failing.

stokito commented:
Hi again, I added a small performance improvement and tried to fix tests. But I can't guarantee that tests are working.
The problem is with line endings of test data files. I using Linux and after checkout all test files have LF line ending.
That's why I always see the same diff error when expected `GET /342 HTTP/1.1[LF]` but actual is `GET /342 HTTP/1.1[CR][LF]`.

More interesting problem with test345. On my computer it is failed because in fact the `tests/log` folder it created two files: `etag345` and `etag345 ` - with space in the end. That's why diff also failed.

If you can check tests on Linux that would be great.
Anyway I checked and retested manually and works fine.

stokito commented:
I just found #5610 and it turned out that Azure CDN may return wrong ETag without quotes. So that logic that I implemented when end was determined by a double quote may cause problems.
I changed logic to just strip CRLF, LF or CR in the end.
Now I thing this branch I ready to go

bagder commented on tests/data/test344:
```diff
@@ -45,14 +45,13 @@ http://%HOSTIP:%HTTPPORT/344 --etag-compare log/etag344 --etag-save log/etag344
+GET /344 HTTP/1.1
+Host: %HOSTIP:%HTTPPORT
+Accept: */*
+
```
I think you removed the CRLF line endings here? The test 344 (and 345) failures look like it:https://github.com/curl/curl/pull/5700/checks?check_run_id=990398257#step:7:1446

stokito replied:
damn, probably this was accidentally changed. Can you fix it? I'm too busy

stokito replied:
ok, I manually switched the file to CRLF

bagder commented:
This would be a faster process if you could look at and address the CI problems yourself. Build problems:
~~~
tool_cb_hdr.c: In function ‘tool_header_cb’:
tool_cb_hdr.c:125:7: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
  125 |       const char *etag_end;
      |       ^~~~~
tool_cb_hdr.c:138:7: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
  138 |       size_t etag_length = (size_t)etag_end - (size_t)etag_start;
      |       ^~~~~~

stokito commented:
please fix CR problem in test yourself, because on my computer git anyway changes them on commit

bagder commented on src/tool_cb_hdr.c:
```diff
@@ -103,14 +116,26 @@ size_t tool_header_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
+      }
+      else {
+        etag_end = skip_eol_at_end(end);
+      }
```
Now this code looks really complicated and hard to follow. Isn't this just finding the last byte of the line that isn't a CR or LF?

stokito replied:
yes, the skip_eol_at_end have a comment. In fact it would be great to extract it to some string utilities

bagder commented on src/tool_cb_hdr.c:
```diff
@@ -103,14 +116,26 @@ size_t tool_header_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
     /* match only header that start with etag (case insensitive) */
     if(curl_strnequal(str, "etag:", 5)) {
-      char *etag_start = ptr + strlen("ETag:");
+      const char *etag_start = ptr + strlen("ETag:");
```
I think you could use '5' instead of a strlen() call here since 5 is used in the line above already. (Then we can ignore the odd fact that the two etag strings are specified using different casings.) Perhaps just:
~~~c
const char *etag_start = &ptr[5];

stokito replied:
See https://github.com/curl/curl/pull/5832

bagder commented on src/tool_cb_hdr.c:
```diff
@@ -103,14 +116,26 @@ size_t tool_header_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
       while(*etag_start && ISSPACE(*etag_start))
         etag_start++;
 
+      /* We expect CRLF as EOL but when called from tests it may be CR or LF */
```
This isn't a test thing. This is how HTTP works in the wild. And it can't be a CR-only.

stokito replied:
never saw this in a wild but guess you already had troubles with this :) The comment may be not so precise but RFC requires CRLF so for simplicity this comment may be fine. If you would like to change then please do because anyway you can write it better.

bagder replied:
I've told you this **many** times already in this PR.

bagder reviewed (commented):
Can you also please squash your commits into maybe one code commit and one test commit, so that reviewing this as a whole gets easier? There's quite a few delta-commits now which makes it harder to view it as a whole.

stokito commented:
You can see the whole diff here https://github.com/curl/curl/pull/5700/files
Small attomic commits are beter for a history when you want to make git blame.

bagder commented:
No, we do not need "small incremental commits" that just show you fixing up problems in the PR from before it was merged. We never do that and we won't start with it now.

stokito commented:
I fixed your review commits, you can press [Squash and Mrrge](https://cloudfour.com/thinks/squashing-your-pull-requests/) and all commits will be reduced to one.

bagder commented:
Sorry, but I found this PR to be too much work and trouble as every single step I need to motivate and put in a lot of work here and still my requests are not adhered to. I've instead created #5833 as an alternative and simpler approach to support weak and other-format etags. I believe that's a better approach then this PR.
